builtin/mainmenu/dlg_create_world.lua

```diff
@@ -51,46 +154,172 @@ local function create_world_formspec(dialogdata)
 				end
 			end
 		end
+
+		local ds = (gameconfig:get("disallowed_settings") or ""):split()
+		for _, value in pairs(ds) do
+			disallowed_settings[value:trim()] = true
+		end
 	end
 
 	local mglist = ""
-	local selindex = 1
+	local selindex
 	local i = 1
+	local first_mg
 	for k,v in pairs(mapgens) do
+		if not first_mg then
+			first_mg = v
+		end
 		if current_mg == v then
 			selindex = i
 		end
 		i = i + 1
 		mglist = mglist .. v .. ","
 	end
+	if not selindex then
+		selindex = 1
+		current_mg = first_mg
+	end
 	mglist = mglist:sub(1, -2)
 
+	local mg_main_flags = function(mapgen, y)
+		if mapgen == "singlenode" then
+			return "", y
+		end
+		if disallowed_settings["mg_flags"] then
+			return "", y
+		end
+
+		local form = "checkbox[4.2," .. y .. ";flag_mg_caves;" ..
+			fgettext("Caves") .. ";"..chflag(flags.main, "caves").."]"
+		y = y + 0.5
+		form = form .. "checkbox[4.2,"..y..";flag_mg_dungeons;" ..
+			fgettext("Dungeons") .. ";"..chflag(flags.main, "dungeons").."]"
+		y = y + 0.5
+
+		if mapgen == "v6" then
```

Comment:
No need to split this into 2 and treat v6 specially due to trees and junglegrass being controlled by flags. The flags already show that trees and junglegrass are controlled by flags.
V6 is only a legacy and old world supporting mapgen, so it is better to keep this code simple and focussed on the non-v6 mapgens.